File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList();
        backtrack(nums, target, ans, cur, 0);
        return ans;
    }

    public void backtrack(
        int[] nums,
        int target,
        List<List<Integer>> ans,
        List<Integer> cur,
        int index
    ) {
        if (target == 0) {
            ans.add(new ArrayList(cur));
        } else if (target < 0 || index >= nums.length) {
            return;
        } else {
            cur.add(nums[index]);
            backtrack(nums, target - nums[index], ans, cur, index);

            cur.remove(cur.get(cur.size() - 1));
            backtrack(nums, target, ans, cur, index + 1);
        }
    }
}
//on^t,ot/n

File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    # 标准的二分搜索框架，搜索目标元素的索引，若不存在则返回 -1
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1   # 注意：右边界是 len(nums)-1

        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1   # 注意：搜索右半区
            else:
                right = mid - 1  # 注意：搜索左半区
        
        return -1
